### DISCUSSION TITLE
Denoising and creating 'mis2mean' maps with preserved subgrain boundaries

### DISCUSSION THREAD

**Discussion by rtaufner** [mtex-toolbox/mtex#613] [Ask Anything]
Hi Mtexers, hopefully everyone is safe and sound.

I have an EBSD analysis containing a porphyroclast and recrystallised grains of Clinopyroxene surrounding it. The 'mis2mean' derived map from it looks good but it appears to have many 'grains within grains'. Most of these small grains within grains are of the same phase as the big grains. See below. I tried to use some demonising filters, such as The Halfquadratic Filter, The Kuwahara Filter, even filling missing data in orientation maps, but all of them smooth away the subgrain boundaries. 

Is there a way to denoise these 'grains within grains' and preserve the sub grain boundaries? The red lines in the figure below are actually the sub grain boundaries, but they look a little odd.

![Intragrain_Misorientation](https://user-images.githubusercontent.com/75056142/101070513-39fb8800-357a-11eb-9421-4cc0ef1832b1.jpg)

**Comment by kilir** [comment]
Hi Rhander,
the halfQuadraticFilter should actually take care of that - preserving LABs while getting rid of noise. Could you give an example (code and image, small crop) of what does not work for you? Many of the red lines look rather like related to misindexed points, right?
Cheers,
Rüdiger

**Reply by rtaufner** [reply]
Hi Rüdiger 
Indeed, some of the red lines are related to misindexed points but many of them are not. See below some examples of the Median, Half Quadratic and Splite filters (in this order). The Splite filter removes all the sub grains. Although the Median and Half Quadradic filters create similar maps, I think that the Median is a bit more 'realistic'. But even so, many sub grains are swept away, although we can still see that the big grain (top right) and small grains (center to lower left) have many internal deformation. 

Which filter would you rather recommend? 

Rhander.

![Median_Filter](https://user-images.githubusercontent.com/75056142/101191465-78f01300-3638-11eb-954d-e3d23d9596a5.jpg)
`%% Median Filter`
`F = medianFilter;`

`% define the size of the window to be used for finding the median`
`F.numNeighbours = 1; `

`% smooth the data`
`ebsdS = smooth(ebsd,F);`

![Half_Quadratic](https://user-images.githubusercontent.com/75056142/101193727-c752e100-363b-11eb-8437-03153810d527.jpg)
`%% Half Quadratic Filter`
`[grains,ebsd('indexed').grainId,ebsd.mis2mean] = calcGrains(ebsd('indexed'));`
`ebsd(grains(grains.grainSize<=3)) = [];`
`[grains,ebsd('indexed').grainId,ebsd.mis2mean] = calcGrains(ebsd('indexed'),'angle',10*degree);`

`F = halfQuadraticFilter;`

`ebsdS = smooth(ebsd('indexed'),F,'fill',grains);`

![Splite_Filter](https://user-images.githubusercontent.com/75056142/101191522-873e2f00-3638-11eb-9c7a-69801fedef88.jpg)
`%% SPLINE FILTER`

`[grains,ebsd('indexed').grainId,ebsd.mis2mean] = calcGrains(ebsd('indexed'));`
`ebsd(grains(grains.grainSize<=3)) = [];`
`[grains,ebsd('indexed').grainId,ebsd.mis2mean] = calcGrains(ebsd('indexed'));`

`F = splineFilter;`
`ebsd = smooth(ebsd('indexed'),F,'fill',grains);`
`[grains,ebsd('indexed').grainId,ebsd.mis2mean] = calcGrains(ebsd('indexed'));`

**Comment by kilir** [comment]
Hi Rhander,

the splineFilter does not preserve edges, hence it's not suprising and and a median (as also the Kuwahara) filter will/can actually create edges since it's a moving window filter. I'd recommend the halfQuadraticFilter with the default `l1dataFit=1` and `l1TV=1` but you may obviously have to adapt the smoothing parameter alpha as well as the threshold angle, which defaults to 15 degree.

Cheers,
Rüdiger

**Reply by rtaufner** [reply]
Hi Rüdiger, 

Thanks for the help. I'll try to run the script with different alpha parameter using the halfQuadraticFilter. 

Cheers, 
Rhander.
